File: src/py/core/game.py

from .game_object_type import game_object_type
from .game_object import game_object
from .cardworld import cardworld
from common import player_location as pl , event as e , card_attribute as ca
from functools import partial
from common import location as loc

class game:

	def __init__ ( self , players , event ):
		self._player = game_object_type()
		for location in dir(pl):
			if location.startswith('__'): continue
			self._player [location] = (location,)

		self._players = dict()

		for i in range(players):
			player = game_object ( self._player )
			self._players [ player.id() ] = player

		self._cardworld = cardworld()

		for player in self._players.values():
			for location in dir(pl):
				if location.startswith('__'): continue
				self._cardworld [ self.location ( player , location) ]

		self._event = event

		event.register_event ( e.reshuffle , self.reshuffle )
		event.register_event ( e.move_card , self.move_card )
		event.register_event ( e.reveal , self.reveal )
		event.register_event ( e.play , self.play )
		event.register_event ( e.cleanup_cardplay , self.cleanup_cardplay )
		event.register_event ( e.discard , self.discard )

	def players ( self ):
		return self._players.values()

	def load_deck ( self , player , deck ):
		for location , cards in deck.items():
			self._cardworld [ self.location ( player , location ) ].stack ( cards )

	def location ( self , player , location ):
		return cardworld.location ( ( loc.player , str(player.id()) ) , player[location] )

	def cardworld ( self ):
		return self._cardworld

	def move_card ( self , origin , destination ):
		origin = cardworld.card_location ( origin.stack() , self.cardworld() [ origin.stack() ].cardstack()._to_key ( origin.card() ) )
		self._event ( e.pre_card_move , origin , destination )
		destination = self._cardworld.move ( origin , destination )
		self._event ( e.post_card_move , origin , destination )

	def _move_top ( self , origin , destination ):
		if len ( self._cardworld [ origin ].cardstack() ) > 0:
			self._event (
				e.move_card,
				cardworld.card_location (
					origin,
					0
				),
				destination
			)
			return False
		return True

	def _move_location ( self , origin , destination ):

		while len ( self._cardworld [ origin ].cardstack() ) > 0:
			self._move_top ( origin , destination )

	def reshuffle ( self , origin , destination ):
		self._move_location ( origin , destination )
		self._cardworld [ destination ].cardstack().shuffle()

	def _move_top_reshuffling ( self , origin , destination , reshuffle_origin ):
		if not self._move_top ( origin , destination ):
			return False

		self._event ( e.reshuffle , reshuffle_origin , origin )

		return self._move_top ( origin , destination )

	def draw ( self , player ):
		location = partial ( self.location , player )

		return self._move_top_reshuffling (
			location ( pl.deck ),
			location ( pl.hand ),
			location ( pl.discard )
		)

	def reveal ( self , player ):
		location = partial ( self.location , player )

		return self._move_top_reshuffling (
			location ( pl.deck ),
			location ( pl.reveal ),
			location ( pl.discard )
		)

	def play ( self , player , card_location ):
		card = self.cardworld() [ card_location ]
		self._event ( e.move_card , card_location , self.location ( player , pl.play ) )

		for i in range ( card [ ca.check_size ] ):
			self._event ( e.reveal , player )

		self.draw ( player )

		self._event ( e.post_play )

	def cleanup_cardplay ( self , player ):
		location = partial ( self.location , player )

		discard = location ( pl.discard )
		self._move_location ( location ( pl.play ) , discard )
		self._move_location ( location ( pl.reveal ) , discard )

	def discard ( self , player , card_location ):
		self.move_card ( card_location , self.location ( player , pl.discard ) )

File: src/cpp/game/src/game.cpp

#include "game.hpp"

#include "common/iterator_range.hpp"

#include "card_attribute.hpp"

namespace game {

namespace stdp = std::placeholders;

#ifdef PSTACK
	static_assert ( false );
#endif

#define PSTACK(l) stack_location ( player , player_location::l )

game::game ( event_manager_type & event_manager , count_type const & n_players , rng_type & rng )
: _players ()
, _base_player ()
, _cardworld ()
, _event_manager ( event_manager )
, _rng ( rng )
{
	register_events();
	create_players ( n_players );
	create_locations ();
}

auto game::stack_location ( object_type const & player , player_location const & loc ) const -> stack_location_type 
{
	if ( ! _players.count ( player.id() ) ) throw 1;

	return stack_location_type ( { location::player , std::to_string ( player.id() ) , loc } );
}

auto game::cardworld () const -> cardworld_type const &
{
	return _cardworld;
}

template < typename T >
auto game::_move ( card_location_type<T> const & origin , stack_location_type const & destination ) -> void
{
	_cardworld.move ( origin , destination );
}

template auto game::_move <game::cardworld_type::key_type> ( card_location_type<cardworld_type::key_type> const & origin , stack_location_type const & destination ) -> void;
template auto game::_move <game::cardworld_type::order_type> ( card_location_type<cardworld_type::order_type> const & origin , stack_location_type const & destination ) -> void;

auto game::move ( any_card_location_type const & origin , stack_location_type const & destination ) -> void
{
	_event_manager.operator() < event_name::pre_move_card > ( std::cref(origin) , std::cref(destination) );

	if (origin.valid_key())
		_move ( origin.key() , destination );
	else
		_move ( origin.order() , destination );

	_event_manager.operator() < event_name::post_move_card > ( std::cref(origin) , std::cref(destination));
}

auto game::reveal ( object_type const & player ) -> void
{
	move_top_reshuffling ( PSTACK ( deck ) , PSTACK ( reveal ) , PSTACK ( discard ) );
}

auto game::play ( object_type const & player , any_card_location_type const & card ) -> void
{
	_event_manager.operator()< event_name::pre_play > ( std::cref(player) , std::cref(card) );

	move ( card , PSTACK ( play ) );

	for ( auto i = _cardworld [ card ] [ card_attribute::successes ]; i > 0; ++i )
		reveal ( player );

	_event_manager.operator()< event_name::post_play > ( std::cref(player) , std::cref(card) );
}

auto game::cleanup_cardplay ( object_type const & player ) -> void
{
	move_location ( PSTACK ( play ) , PSTACK ( discard ) );
	move_location ( PSTACK ( reveal ) , PSTACK ( discard ) );
}

auto game::discard ( object_type const & player , any_card_location_type const & card ) -> void
{
	move ( card , PSTACK ( discard ) );
}

auto game::move_top ( stack_location_type const & origin , stack_location_type const & destination ) -> bool
{
	if ( _cardworld [ origin ].size() > 0 ) {
		move ( order_card_location_type ( origin , 0 ) , destination );
		return false;
	}
	return true;
}

auto game::move_location ( stack_location_type const & origin , stack_location_type const & destination ) -> void
{
	auto & origin_stack = _cardworld [ origin ];
	while ( origin_stack.size() > 0 )
		move_top ( origin , destination );
}

auto game::register_events () -> this_type &
{
	_event_manager.register_event<event_name::reshuffle> ( std::bind ( &game::reshuffle , this , stdp::_1 , stdp::_2 ) );
	_event_manager.register_event<event_name::move_card> ( std::bind ( &game::move , this , stdp::_1 , stdp::_2 ) );
	_event_manager.register_event<event_name::reveal> ( std::bind ( &game::reveal , this , stdp::_1 ) );
	_event_manager.register_event<event_name::play> ( std::bind ( &game::play , this , stdp::_1, stdp::_2 ) );
	_event_manager.register_event<event_name::cleanup_cardplay> ( std::bind ( &game::cleanup_cardplay , this , stdp::_1 ) );
	_event_manager.register_event<event_name::discard> ( std::bind ( &game::discard , this , stdp::_1 , stdp::_2 ) );

	return *this;
}

auto game::create_players ( count_type const & n_players ) -> this_type &
{
	for ( auto i = n_players; i > 0; --i ) {
		object_type player ( _base_player );
		_players.insert ( std::make_pair ( player.id() , std::move ( player ) ) );
	}
	return *this;
}

auto game::create_locations () -> this_type &
{
	for ( auto const & player : _players ) {
		for ( auto const & loc : common::make_range ( player_location::begin() , player_location::end() ) ) {
			_cardworld [ stack_location ( player.second , loc ) ];
		}
	}

	return *this;
}

auto game::reshuffle ( stack_location_type const & origin , stack_location_type const & destination ) -> void
{
	move_location ( origin , destination );
	_cardworld [ destination ].shuffle ( _rng );
}

auto game::move_top_reshuffling ( stack_location_type const & origin , stack_location_type const & destination , stack_location_type const & reshuffle_origin ) -> bool
{
	if ( ! move_top ( origin , destination ) ) return false;

	reshuffle ( reshuffle_origin , origin );

	return move_top ( origin , destination );
}

#undef PSTACK

}

